src/ur/workday.ur

```diff
@@ -0,0 +1,378 @@
+                (* TODO: maybe this should return option worker? *)
+
+            val getAll : transaction (list $worker) =
+                depaginate 100 Common "/workers"
+        end
+
+        structure FeedbackApi = struct
+            fun get (workerId : wid) : transaction (list anytimeFeedback) =
+                depaginate 100 PerformanceEnablement ("/workers/" ^ workerId ^ "/anytimeFeedbackEvents")
+                (*return (
+                  (* Filter out feedback where the provider isn't in the workers table. *)
+                  (* HACK *)
+                  (* TODO: can we guarantee this doesn't happen? *)
+                  (*List.filter
+                    (fn fb => List.mem (Option.getOrError <xml>no</xml> fb.FromWorker).Id workers)*)
+                    (fromJson s : response (list anytimeFeedback)).Data)*)
+
+            fun getAll (workers : list wid) : transaction (list anytimeFeedback) =
+                List.mapConcatM get workers
+
+            fun parseSoap (xmlString : string) : wid =
+                (* HACK: actually parse XML *)
+                case String.ssplit {Haystack = xmlString, Needle = "<wd:ID wd:type=\"WID\">"} of
+                    None => error <xml>Could not find ID in response: {[xmlString]}</xml>
+                  | Some (pre, post) =>
+                    case String.split post #"<" of
+                        None => error <xml>Badly formed response: {[xmlString]}</xml>
+                      | Some (pre, post) => pre
+            fun post (request : $feedbackWithoutId) : transaction wid =
+                let fun toSoapXml request =
+                        (* FIXME: escape user-provided values *)
+                        "<?xml version=\"1.0\" encoding=\"utf-8\"?>
+<env:Envelope xmlns:env=\"http://schemas.xmlsoap.org/soap/envelope/\">
+    <env:Body>
+        <bsvc:Give_Feedback_Request xmlns:bsvc=\"urn:com.workday/bsvc\" bsvc:version=\"v42.2\">
+            <bsvc:Give_Feedback_Data>
+                <bsvc:From_Worker_Reference>
+                    <bsvc:ID bsvc:type=\"WID\">" ^ request.Provider ^ "</bsvc:ID>
+                </bsvc:From_Worker_Reference>
+                <bsvc:To_Workers_Reference>
+                    <bsvc:ID bsvc:type=\"WID\">" ^ request.About ^ "</bsvc:ID>
+                </bsvc:To_Workers_Reference>
+                <bsvc:Comment>" ^ request.Body ^ "</bsvc:Comment>
+                <bsvc:Show_Name>true</bsvc:Show_Name>
+                <bsvc:Confidential>true</bsvc:Confidential>
+            </bsvc:Give_Feedback_Data>
+        </bsvc:Give_Feedback_Request>
+    </env:Body>
+</env:Envelope>"
+        (* TODO: extra information: status, dates, requestedBy, etc *)
+
+                in
+                  body <- return (toSoapXml request);
+                  tok <- token;
+                  debug tok;
+                  url <- return "https://wcpdev-services1.wd101.myworkday.com/ccx/service/janestreet_wcpdev1/Talent/v42.2";
```

Comment:
Surely a common URL prefix can be reused within the file, rather than copied for each different usage.

Reply:
Yeah this one is a bit annoying since it's formatted slightly differently to the REST API URLs, but I made a change that reflects a best-effort